### PR TITLE
Handle posts > HPOS redirect more gracefully when backup post is missing

### LINKED ISSUE #44177: Missing HPOS placeholder posts not handled gracefully
**Prerequisites (mark completed items with an [x]):**
- [x] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [x] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

**Issue Description:**

When purging the HPOS placeholder posts (post type `shop_order_placehold`) and then visiting `/wp-admin/post.php?post=123&action=edit`, instead of the redirect to the HPOS order `/wp-admin/admin.php?page=wc-orders&action=edit&id=123` WooCommerce fails with a 500 server error. (`123` being an order ID.)

_(Note that purging these posts is not officially supported yet. Once it is, WooCommerce will provide a user interface for purging this data. However, WooCommerce should handle the absence of these posts more gracefully.)_

### PR DESCRIPTION
**Pull request by jorgeatorres**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

When HPOS is authoritative, we automatically redirect admin URLs for order posts to the corresponding HPOS one. For example, `/wp-admin/post.php?post=123&action=edit` gets turned into `/wp-admin/admin.php?page=wc-orders&action=edit&id=123`.

This works correctly when the order post exists, but was producing a 500 error when not.

This is sort of an edge case as we don't endorse removal of the backup post / placeholder post, but we've seen some users doing that after migrating to HPOS anyways.

This PR adds an additional check so that the redirect succeeds in those cases (as long as the HPOS order exists).

Closes #44177.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure HPOS is configured as datastore in WC > Settings > Advanced > Features and that compatibility mode is disabled.
2. Ensure your site has a few orders. If necessary, create some or use [Smooth Generator](https://github.com/woocommerce/wc-smooth-generator) for a smoother testing experience. Make note of some order IDs.
3. Delete the backup post for one of the orders by running the following command:
   ```
   wp db query "DELETE FROM $(wp db prefix)posts WHERE ID = <order_id>"
   ```
4. Visit `https://yoursite.test/wp-admin/post.php?post=<order_id>&action=edit` and confirm that you're redirected to the (HPOS) order edit screen.
5. Visit `https://yoursite.test/wp-admin/post.php?post=<post_id>&action=edit` with a `<post_id>` that has no corresponding post and confirm that you see an error.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Internal/Admin/Orders/PostsRedirectionController.php [hunk 1]
```diff
@@ -130,7 +130,16 @@ private function maybe_redirect_to_edit_order_page(): void {
 		$redirect_from_types   = wc_get_order_types( 'admin-menu' );
 		$redirect_from_types[] = 'shop_order_placehold';
 
-		if ( ! $post_id || ! in_array( get_post_type( $post_id ), $redirect_from_types, true ) || ! isset( $_GET['action'] ) ) {
+		$order_type = get_post_type( $post_id );
+		if ( ! $order_type ) {
+			// Attempt to determine whether this is a valid order on the HPOS side before bailing out.
+			$datastore = new \WC_Data_Store( 'order' );
```
Comment (vedanshujain) [inline plugins/woocommerce/src/Internal/Admin/Orders/PostsRedirectionController.php]: should this be `\WC_Data_Store::load( 'order' );`. It's the same things, but seems like we use this more often than the `new` keyword.

Comment (vedanshujain) [reply plugins/woocommerce/src/Internal/Admin/Orders/PostsRedirectionController.php]: [optional] Also maybe we can use `OrderUtil::is_order` or `OrderUtil::get_order_type` here? That way we will be using methods that are covered by unit tests, wdyt?

**Review by vedanshujain — COMMENTED** [review]
left couple of minor comments, lgtm otherwise

File: plugins/woocommerce/src/Internal/Admin/Orders/PostsRedirectionController.php (hunk 1 above)
Comment (jorgeatorres) [reply plugins/woocommerce/src/Internal/Admin/Orders/PostsRedirectionController.php]: That's a good idea. Thanks!

**Comment by jorgeatorres** [comment]
@vedanshujain: Thanks for the review! I've implemented your suggestion and simplified the logic a bit. Please let me know what you think.

**Review by vedanshujain — APPROVED** [review]
nice, LGTM!

**Pull request merged**